### DISCUSSION TITLE
Is it possible to use a compound key (two or more keys) as a predicate for `updateOrCreateMany`?

### DISCUSSION THREAD

**Discussion by vtduong** [adonisjs/core#4368] [Help]
I would like to update many rows based on a compound key or where two or more conditions are met, something like:
```typescript
Address.updateOrCreateMany(['city', 'country'], addresses)
```
Is it possible in Adonisjs v5 and v6?

In Adonisjs v5, the `predicate` parameter of the method accepts a list of keys but only the first key is used in the WHERE clause of the executed statement.

I appreciate any help!

**Comment by thetutlage** [comment]
Can you please enable query debugging and share the SQL queries executed by the `updateOrCreateMany` method?

**Reply by vtduong** [reply]
Thank you for your reply! Sure thing!
For context, I'm still using Adonisjs v5, 
Below is the Typescript code that calls `updateOrCreateMany`
```typescript
// 'id' and 'version' are both defined as primary keys of the table in my db schema as followed:
// table.primary(['id', 'version'])
Opportunity.updateOrCreateMany(['id', 'version'], opps)
```

Here's the debugging logs 
```
"pg" Opportunity (17 ms) UPDATE "opportunities" SET "updated_at" = ?, "form_jsonb" = ? WHERE "id" = ? [
  '2024-01-30T22:39:41.162-05:00',
  '{"fund_id":"7efd797a-9878-4da3-a54c-d7330286eabf","some_field":"abc","vc_firm_id":"e0505cc1-f918-4e66-9867-f18c6193e8cf","opportunity_id":"249b160f-8b8c-4404-9b8e-0be6d7be8d3b","opportunity_name":"compA opp v2","current_stage_idx":0,"current_status_id":"cd92798c-71fb-476c-a29f-20408e0a0376","starting_stage_idx":0,"primary_team_lead_id":"e7a256ed-ff79-4b5a-843e-d509f31aa7a0","field_2":null,"field_3":null}',
  '249b160f-8b8c-4404-9b8e-0be6d7be8d3b'
]
"pg" Opportunity (6.87 ms) UPDATE "opportunities" SET "updated_at" = ?, "form_jsonb" = ? WHERE "id" = ? [
  '2024-01-30T22:39:41.181-05:00',
  '{"fund_id":"7efd797a-9878-4da3-a54c-d7330286eabf","some_field":"abc","vc_firm_id":"e0505cc1-f918-4e66-9867-f18c6193e8cf","opportunity_id":"56e37023-9abf-4978-b27d-c1266cbf5a88","opportunity_name":"compB opp v2","current_stage_idx":0,"current_status_id":"cd92798c-71fb-476c-a29f-20408e0a0376","starting_stage_idx":0,"primary_team_lead_id":"e7a256ed-ff79-4b5a-843e-d509f31aa7a0","field_2":null,"field_3":null}',
  '56e37023-9abf-4978-b27d-c1266cbf5a88'
]
```
I intend to migrate to v6 soon but based on the [docs](https://lucid.adonisjs.com/docs/crud-operations#updateorcreatemany), it seems not possible to pass a composite key as predicate to the method

**Reply by thetutlage** [reply]
How come there are select queries in the logs?

Also, the latest version of Lucid supports everything from the last original (used with v5).

**Reply by vtduong** [reply]
The logs were generated from my test case where I have some select statements before and after `updateOrCreateMany()`
Does it support composite keys as predicate to search for records to be updated?

**Reply by thetutlage** [reply]
Yes, it does. But, since you have not shared the full logs (select queries are missing), I cannot tell why it does not work in your case

**Reply by vtduong** [reply]
I apologize for not including the full logs as it's quite lengthy (most queries are for setting up test data) and I thought it wasn't relevant to the issue. Please see the attached file:
[test-log.log](https://github.com/adonisjs/core/files/14160750/test-log.log)

I really appreciate your time and support!

**Reply by thetutlage** [reply]
Well, all I want is to isolate this issue and share its log. Either a log file of everything and neither the incomplete logs. Or you can create a new repo, reproduce the issue and share that with me

**Reply by vtduong** [reply]
Thank you for your guidance! here's the relevant log resulted from executing the below code
```typescript
Opportunity.updateOrCreateMany(['id', 'version'], opps)
```
![image](https://github.com/adonisjs/core/assets/14185568/91f1f079-a2f4-4307-9335-bccee7b496cf)

From the screenshot above, the SELECT FOR UPDATE statement does include both `id` and `version` as part of  the WHERE clause but the UPDATE statement includes only the `id` which updates all rows with that `id` regardless of `version`. Is this the expected behavior of `updateOrCreateMany`?

**Reply by thetutlage** [reply]
I think, the update clause uses the `primaryKey` from the model, which in this case is the `id` column. So, if the primarKey alone is not unique, then you might run into other issues as well beyond `updateOrCreateMany` method.

This is because, Lucid does not support compound primary keys. Again, `updateOrCreateMany` can pass multiple keys for the `SELECT` query, but update will always be done using the `primaryKey`.

**Reply by vtduong** [reply]
I see, thank you for your detailed explanation! I found a [discussion](https://github.com/adonisjs/lucid/pull/382) about this topic but it dated 2019 so I wonder if Lucid supports primary composite key in Adonis v6?

**Reply by RoncoCo** [reply]
@thetutlage when you say "Lucid does not support compound primary keys" you mean that models that have "isPrimary: true" on more than one field will not have the desired behavior when executing CreateOrUpdate? meaning that not only the record which matches all the PKs will be updated?

**Reply by thetutlage** [reply]
@RoncoCo Yes. Exactly that!
